fusillade/api/oauth.py

```diff
@@ -0,0 +1,182 @@
+    client_id = query_params.get("client_id")
+    client_id = client_id if client_id != 'None' else None
+    if client_id:
+        # TODO: audit this
+        auth_params = dict(client_id=query_params["client_id"],
+                           response_type="code",
+                           scope=query_params["scope"],
+                           redirect_uri=query_params["redirect_uri"],
+                           state=query_params["state"])
+        if "audience" in query_params:
+            auth_params["audience"] = query_params["audience"]
+    else:
+        state = base64.b64encode(json.dumps(query_params).encode()).decode()
+        # TODO: set random state
+        # openid_provider = query_params["openid_provider"]
+        oauth2_config = Config.get_oauth2_config()
+        auth_params = dict(client_id=oauth2_config[openid_provider]["client_id"],
+                           response_type="code",
+                           scope="openid email profile",
+                           redirect_uri=oauth2_config[openid_provider]["redirect_uri"],
+                           state=state)
+    dest = furl(get_openid_config(openid_provider)["authorization_endpoint"]).add(query_params=auth_params).url
+    return ConnexionResponse(status_code=requests.codes.found, headers=dict(Location=dest))
+
+
+def proxy_response(dest_url, method='GET', headers=None, body='', query_params=None, **extra_query_params):
+    if query_params or extra_query_params:
+        dest_url = furl(dest_url).add(dict(query_params, **extra_query_params)).url
+    proxy_res = requests.request(method=method,
+                                 url=dest_url,
+                                 headers=headers,
+                                 data=body)
+    if proxy_res.headers.get("Content-Type", "").startswith("application/json"):
+        body = proxy_res.json()
+    else:
+        body = proxy_res.content
+    for header in "connection", "content-length", "date":
+        proxy_res.headers.pop(header, None)
+    proxy_resp = ConnexionResponse(body=body, status_code=proxy_res.status_code, headers=dict(proxy_res.headers))
+    return proxy_resp
+
+
+def serve_openid_config():
+    '''
+     Part of OIDC
+    '''
+    openid_config = get_openid_config(os.environ["OPENID_PROVIDER"])
+    auth_host = request.headers['host']
+    # TODO figure out of auth_host should be os.environ["API_DOMAIN_NAME"]
+    openid_config.update(authorization_endpoint=f"https://{auth_host}/authorize",
+                         token_endpoint=f"https://{auth_host}/oauth/token",
+                         jwks_uri=f"https://{auth_host}/.well-known/jwks.json",
+                         revocation_endpoint=f"https://{auth_host}/oauth/revoke",
+                         userinfo_endpoint=f"https://{auth_host}/userinfo")
+    return ConnexionResponse(body=openid_config, status_code=requests.codes.ok)
+
+
+def serve_jwks_json():
+    '''
+     Part of OIDC
```

Comment:
One less leading space here and elsewhere

Reply:
Fixed